File: Behavioral/Lab20/Typescript/src/Entities/apartment.ts
import { RealEstate } from './real_estate';
import { Customer } from './customer';
import { Realtor } from './realtor';
export class Apartment extends RealEstate {
    public roomsNumber: number;
    public heatingType: string;

    constructor();
    constructor(address: string, description: string, area: number, price: number, status: string, roomsNumber: number);
    constructor(address?: string, description?: string, area?: number, price?: number, status?: string, roomsNumber?: number) {
        if (address && description && area && price && status && roomsNumber) {
            super(address, description, area, price, status);
            this.roomsNumber = roomsNumber;
        } else {
            super("Dungeon", "tra ta ta", 50, 1000, "New");
            this.roomsNumber = 3;
        }
    }

    public toString(): string {
        return "Apartment:\n" + super.toString() + `, Rooms number: ${this.roomsNumber}`;
    }


    public clone(): Apartment {
        const clone = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
        clone.realtor = this.realtor ? Realtor.newRealtorFromObject(this.realtor) : null;
        return clone;
    }

    public shallowCopy(): Apartment {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }

    public reset(): void {
        super.reset();
        this.roomsNumber = 0;
        this.heatingType = "";
    }
}

File: Behavioral/Lab20/C#/Entities/Apartment.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C_.Entities
{
    public class Apartment : RealEstate
    {
        public int RoomsNumber{ get; set;}
        public string HeatingType { get; set; }
        public Apartment():base("Dungeon", "tra ta ta", 50, 1000, "New"){
            RoomsNumber = 3;
        }
        public Apartment(string address, string description, double area, double price, string status, int roomsNumber) :
            base(address, description, area, price, status)
        {
            RoomsNumber = roomsNumber;
        }
        public override string ToString()
        {
            return "Apartment:\n" + base.ToString() + $", Rooms number: {RoomsNumber}";
        }

        public override Apartment Clone()
        {
            var clone = (Apartment)MemberwiseClone();
            clone.Realtor = Realtor == null ? Realtor : new Realtor(Realtor);
            return clone;
        }

        public override Apartment ShallowCopy()
        {
            return (Apartment)MemberwiseClone();
        }

        public override void Reset()
        {
            base.Reset();
            RoomsNumber = 0;
            HeatingType = "";
        }
    }
}
